README.md

Couchbase store for node cache manager
==================================

Couchbase cache store for [node-cache-manager](https://github.com/BryanDonovan/node-cache-manager). 

Installation
------------

```sh
npm install cache-manager-couchbase --save
```
or
```sh
yarn add cache-manager-couchbase
```

Usage Examples
--------------

TBD

CHANGELOG.md
# Changelog

## v0.1.0 - Feb 2022 
First release


test/index.test.js
const cacheManager = require('cache-manager');
const couchbaseStore = require('../index');
const couchbase = require('couchbase');

let couchbaseCache;
let customCouchbaseCache;

var config = {
  connectionString: 'couchbase://127.0.0.1:8091',
  connectionOptions: {
    username: 'Administrator',
    password: 'couchbase'
  },
  bucket: 'test-bucket',
  scope: '_default',
  collection: '_default',
  ttl: 2
};

beforeAll((done) => {
  couchbaseCache = cacheManager.caching({
    store: couchbaseStore,
    connectionString: config.connectionString,
    connectionOptions: config.connectionOptions,
    bucket: config.bucket,
    scope: config.scope,
    collection: config.collection,
    ttl: config.ttl,
  });

  done();
});

afterAll(async () => {
  await couchbaseCache.store.getClient().close();
});

describe('set', () => {
  it('should return a MutationResult object when succesful', async () => {
    const res = await couchbaseCache.set('foo', 'bar');
    expect(res).not.toBeUndefined();
  });

  it('should store a value without ttl', (done) => {
    couchbaseCache.set('foo', 'bar', (err) => {
      expect(err).toEqual(null);
      done();
    });
  });

  it('should store a value with a specific ttl', (done) => {
    couchbaseCache
      .set('foo', 'bar', { ttl: 60 }, (err) => {
        expect(err).toEqual(null);
        done();
      });
  });


});

describe('get', () => {
  it('should return a promise', (done) => {
    expect(couchbaseCache.get('foo')).toBeInstanceOf(Promise);
    done();
  });

  it('should resolve promise on success', async () => {

    const setRes = await couchbaseCache.set('foo', 'bar');
    const getRes = await couchbaseCache.get('foo');

    expect(getRes).toEqual('bar');
  });

  /*
  // this causes a segmentation fault
  it('should reject promise on error', async () => {

    try {
      let result = await couchbaseCache.get('foo', () => new Error('Something went wrong'));
    } catch (err) {
      expect(err.message).toEqual('Something went wrong');
    }
  });*/
});


spec/CacheManagerCouchbaseSpec.js
const cacheManager = require('cache-manager');
const couchbaseStore = require('../index');
const couchbase = require('couchbase');

let couchbaseCache;
let customCouchbaseCache;

var config = {
  connectionString: 'couchbase://127.0.0.1:8091',
  connectionOptions: {
    username: 'Administrator',
    password: 'Administrator'
  },
  bucket: 'test-bucket',
  scope: '_default',
  collection: '_default',
  ttl: 2
};

beforeAll(async () => {
  couchbaseCache = cacheManager.caching({
    connectionString: config.connectionString,
    store: couchbaseStore,
    connectionOptions: config.connectionOptions,
    scope: config.scope,
    bucket: config.bucket,
    collection: config.collection,
    ttl: config.ttl
  });
  
  await couchbaseCache.reset();
});

afterAll(async () => {
  await couchbaseCache.store.getClient().close();
});



describe('set', () => {
  it('should return a MutationResult object when succesful', async () => {
    const res = await couchbaseCache.set('foo', 'bar');
    expect(res).not.toBeUndefined();
  });

  it('should store a value without ttl', (done) => {
    couchbaseCache.set('foo', 'bar', (err) => {
      expect(err).toEqual(null);
      done();
    });
  });

  it('should store a value with a specific ttl', (done) => {
    couchbaseCache
      .set('foo', 'bar', { ttl: 60 }, (err) => {
        expect(err).toEqual(null);
        done();
      });
  });

  it('should not store an invalid value', (done) => {
    couchbaseCache.set('foo1', undefined, (err) => {
      try {
        expect(err).not.toEqual(null);
        expect(err.message).toEqual('"undefined" is not a cacheable value');
        done();
      } catch (e) {
        done(e);
      }
    });
  });

  it('should reject promise on error', (done) => {
    couchbaseCache.set('foo', null)
      .then(() => done(new Error('Should reject')))
      .catch(() => done());
  });

  it('should not be able to store a null value (not cacheable)', (done) => {
    couchbaseCache.set('foo2', null, (err) => {
      if (err) {
        return done();
      }
      done(new Error('Null is not a valid value!'));
    });
  });

  it('should store a value without callback', (done) => {
    couchbaseCache.set('foo', 'baz');
    couchbaseCache.get('foo', (err, value) => {
      expect(err).toEqual(null);
      expect(value).toEqual('baz');
      done();
    });
  });

});

describe('get', () => {
  it('should return a promise', (done) => {
    expect(couchbaseCache.get('foo')).toBeInstanceOf(Promise);
    done();
  });

  it('should resolve promise on success', (done) => {
    couchbaseCache.set('foo', 'bar')
      .then(() => couchbaseCache.get('foo'))
      .then(result => {
        expect(result).toEqual('bar');
        done();
      });
  });

  it('should reject promise on error', async () => {
    const client = couchbaseCache.store.getClient();
    client.get = (key, cb) => cb(new Error('Something went wrong'));

    await couchbaseCache.get('foo')
      .catch((err) => {
        expect(err.message).toEqual('Something went wrong');
      })
  });

  it('should retrieve a value for a given key', (done) => {
    const value = 'bar';
    couchbaseCache.set('foo', value, () => {
      couchbaseCache.get('foo', (err, result) => {
        expect(err).toEqual(null);
        expect(result).toEqual(value);
        done();
      });
    });
  });

  it('should retrieve a value for a given key if options provided', async () => {
    const value = 'bar';
    const firstOp = await couchbaseCache.set('foo', value);
    const secondOp = await couchbaseCache.get('foo', {});
    expect(secondOp).toEqual(value);
  });

  it('should return null when the key is invalid', (done) => {
    couchbaseCache.get('invalidKey', (err, result) => {
      expect(err).toEqual(null);
      expect(result).toEqual(null);
      done();
    });
  });
});

describe('del', () => {
  it('should delete a value for a given key', (done) => {
    couchbaseCache.set('foo', 'bar', () => {
      couchbaseCache.del('foo', (err) => {
        expect(err).toEqual(null);
        done();
      });
    });
  });

  it('should delete a value for a given key without callback', (done) => {
    couchbaseCache.set('foo', 'bar', () => {
      couchbaseCache.del('foo');
      done();
    });
  });
});

describe('reset', () => {

  jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000;

  it('should flush underlying db', (done) => {
    couchbaseCache.reset((err) => {
      expect(err).toEqual(null);
      done();
    });
  });

  it('should flush underlying db without callback', (done) => {
    couchbaseCache.reset();
    done();
  });
});

describe('ttl', () => {
  it('should retrieve ttl for a given key', async () => {

    await couchbaseCache.set('foo', 'bar');
    await couchbaseCache.ttl('foo', (err, ttl) => {
      expect(err).toEqual(null);
      expect(ttl).not.toEqual(null);
    });
  });


  it('should retrieve ttl for an invalid key', async () => {
    await couchbaseCache.ttl('invalidKey', (err, ttl) => {
      expect(err).toEqual(null);
      expect(ttl).not.toEqual(null);
    });
  });
});

describe('keys', () => {
  it('should return a promise', (done) => {
    expect(couchbaseCache.keys('foo')).toBeInstanceOf(Promise);
    done();
  });

  it('should resolve promise on success', (done) => {
    couchbaseCache.set('foo', 'bar')
      .then(() => couchbaseCache.keys('f*'))
      .then(result => {
        expect(result.indexOf('foo')).not.toEqual(-1);
        done();
      });
  });

  it('should return an array of keys for the given pattern', (done) => {
    couchbaseCache.set('foo', 'bar', () => {
      couchbaseCache.keys('f*', (err, arrayOfKeys) => {
        expect(err).toEqual(null);
        expect(arrayOfKeys).not.toEqual(null);
        expect(arrayOfKeys.indexOf('foo')).not.toEqual(-1);
        done();
      });
    });
  });

  it('should return an array of keys without pattern', (done) => {
    couchbaseCache.set('foo', 'bar', () => {
      couchbaseCache.keys((err, arrayOfKeys) => {
        expect(err).toEqual(null);
        expect(arrayOfKeys).not.toEqual(null);
        expect(arrayOfKeys.indexOf('foo')).not.toEqual(-1);
        done();
      });
    });
  });

});

describe('isCacheableValue', () => {
  it('should return true when the value is not undefined', () => {
    expect(couchbaseCache.store.isCacheableValue(0)).toBe(true);
    expect(couchbaseCache.store.isCacheableValue(100)).toBe(true);
    expect(couchbaseCache.store.isCacheableValue('')).toBe(true);
    expect(couchbaseCache.store.isCacheableValue('test')).toBe(true);
  });

  it('should return false when the value is undefined', () => {
    expect(couchbaseCache.store.isCacheableValue(undefined)).toBe(false);
  });

  it('should return false when the value is null', () => {
    expect(couchbaseCache.store.isCacheableValue(null)).toBe(false);
  });
});

describe('wrap function', () => {
  // Simulate retrieving a user from a database
  function getUser(id, cb) {
    setTimeout(() => {
      cb(null, { id: id });
    }, 100);
  }

  // Simulate retrieving a user from a database with Promise
  function getUserPromise(id) {
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve({ id: id });
      }, 100);
    });
  }

  it('should be able to cache objects', (done) => {
    const userId = 123;

    // First call to wrap should run the code
    couchbaseCache.wrap('wrap-user', (cb) => {
      getUser(userId, cb);
    }, (err, user) => {
      expect(user.id).toEqual(userId);

      // Second call to wrap should retrieve from cache
      couchbaseCache.wrap('wrap-user', (cb) => {
        getUser(userId + 1, cb);
      }, (err, user) => {
        expect(user.id).toEqual(userId);
        done();
      });
    });
  });

  it('should work with promises', () => {
    const userId = 123;

    // First call to wrap should run the code
    return couchbaseCache
      .wrap(
        'wrap-promise',
        () => getUserPromise(userId),
      )
      .then((user) => {
        expect(user.id).toEqual(userId);

        // Second call to wrap should retrieve from cache
        return couchbaseCache.wrap(
          'wrap-promise',
          () => getUserPromise(userId + 1),
        )
          .then((user) => expect(user.id).toEqual(userId));
      });
  });
});


index.js
const couchbase = require('couchbase');
Promise = require('bluebird');


/* configuration example.. see couchbase docs for more
   * {
   *    connectionString:'',
   *    connectionOptions: {
   *      username:'',
   *      password:''
   *    },
   *    bucket: '',
   *    scope: '',
   *    collection: '',
   *    ttl: 10
   * } 
   */
const couchbaseStore = (...args) => {

    const globalOptions = args[0];
    let cluster = new couchbase.Cluster(globalOptions.connectionString, globalOptions.connectionOptions);
    let couchbaseBucket = cluster.bucket(globalOptions.bucket);
    let couchbaseCollection = null;
    let couchbaseBucketManager = null;
    let globalTtl = 0;
    let isCacheableValue = (value) => ((value !== null && value !== undefined) ? true : false);
    
    if (globalOptions.scope != null && globalOptions.scope !== 'undefined') {
        if (globalOptions.collection != null && globalOptions.collection !== 'undefined') {
            couchbaseCollection = couchbaseBucket.scope(globalOptions.scope).collection(globalOptions.collection);
        }
    } else {
        couchbaseCollection = couchbaseBucket.defaultCollection();
    }

    couchbaseBucketManager = cluster.buckets();

    let self = {
        name: 'couchbase',
        options: globalOptions,
        isCacheableValue
    };

    self.getClient = () => {
        return cluster;
    }

    self.set = (key, value, options, cb) => {

        if (typeof options === 'function') {
            cb = options;
            options = {
                ttl: 0
            };
        }

        if (options == null || options === 'undefined') {
            options = {
                ttl: 0
            };
        }

        if (cb === undefined) {
            return new Promise(function (resolve, reject) {
                self.set(key, value, options, function (err, result) {
                    err ? reject(err) : resolve(result)
                })
            })
        }

        if (!isCacheableValue(value)) {
            return cb(new Error(`"${value}" is not a cacheable value`));
        }

        let ttl = (options.ttl || options.ttl === 0) ? options.ttl : globalTtl;
        ttl = ttl * 60;

        couchbaseCollection.upsert(key, value, { expiry: ttl }, cb);
    };

    self.get = (key, options, cb) => {
        if (typeof options === 'function') {
            cb = options;
        }

        if (cb === undefined) {
            return new Promise(function (resolve, reject) {
                self.get(key, options, function (err, result) {
                    err ? reject(err) : resolve(result)
                })
            })
        }

        couchbaseCollection.get(key)
            .then((result) => {
                return cb(null, result.content);
            })
            .catch((err) => {
                if(err.message == 'document not found') {
                    return cb(null, null);
                }
                return cb(err);
            });
    };

    self.del = (key, options, cb) => {
        if (typeof options === 'function') {
            cb = options;
        }

        if (cb === undefined) {
            return new Promise(function (resolve, reject) {
                self.del(key, options, function (err, result) {
                    err ? reject(err) : resolve(result)
                })
            })
        }

        couchbaseCollection.remove(key, cb);
    };

    self.reset = (cb) => {

        if (cb === undefined) {
            return new Promise(function (resolve, reject) {
                self.reset(function (err, result) {
                    err ? reject(err) : resolve(result)
                })
            })
        }

        return couchbaseBucketManager.flushBucket(globalOptions.bucket, cb);
    };

    self.keys = (pattern, cb) => {

        if (typeof pattern === 'function') {
            cb = pattern;
            pattern = '';
        }

        if (cb === undefined) {
            return new Promise(function (resolve, reject) {
                self.keys(pattern, function (err, result) {
                    err ? reject(err) : resolve(result)
                })
            })
        }

        let bucketReplacement = null;

        if (globalOptions.collection == null && globalOptions.collection === 'undefined') {
            bucketReplacement = ` \`${globalOptions.bucket}\` b `;
        } else {
            bucketReplacement = ` \`${globalOptions.bucket}\`.\`${globalOptions.scope}\`.\`${globalOptions.collection}\` b `;
        }

        const query = `
            SELECT RAW meta(b).id
            FROM ${bucketReplacement} 
            WHERE REGEXP_CONTAINS(META(b).id, $PATTERN)
          `;

        const optionsQuery = {
            parameters: {
                PATTERN: pattern
            }
        }

        cluster.query(query, optionsQuery)
            .then((result) => {
                return cb(null, result.rows);
            })
            .catch((err) => {
                return cb(err);
            });
    };

    self.ttl = (key, cb) => {

        if (cb === undefined) {
            return new Promise(function (resolve, reject) {
                self.ttl(key, function (err, result) {
                    err ? reject(err) : resolve(result)
                })
            })
        }

        couchbaseCollection.get(key)
            .then((result) => {
                return (result.expiryTime - Math.floor(Date.now() / 1000));
            })
            .then((secondsToExpiration) => {
                return cb(null, secondsToExpiration);
            })
            .catch((err) => {
                if(err.message == 'document not found') {
                    return cb(null, -1);
                }
                return cb(err);
            });
    };

    return self;
};

const methods = {
    create: (...args) => couchbaseStore(...args),
};

module.exports = methods;

index_old_old.js
'use strict';

const couchbase = require('couchbase');

/* configuration example.. see couchbase docs for more
   * {
   *    connectionString:'',
   *    connectionOptions: {
   *      username:'',
   *      password:''
   *    },
   *    bucket: '',
   *    scope: '',
   *    collection: '',
   *    ttl: 10
   * } 
   */

class CouchbaseStore {

    constructor(args) {
        var self = this;

        self.connectionString = args.connectionString;
        self.connectionOptions = args.connectionOptions;
        self.bucket = args.bucket;
        self.scope = args.scope;
        self.collection = args.collection;
        self.ttl = args.ttl;
        self.name = 'couchbase';
        self.isCacheableValue = args.isCacheableValue || (value => value !== undefined && value !== null);
    }
    

    async prepCouchbaseConnection() {
        var self = this;
        self.clients= {
            couchbaseBucket: null,
            couchbaseCluster: null,
            couchbaseCollection: null
        };
        self.clients.couchbaseCluster  = await couchbase.connect(self.connectionString, self.connectionOptions);
        self.clients.couchbaseBucket= self.clients.couchbaseCluster.bucket(self.bucket);
        if (self.scope != null && self.scope !== 'undefined') {
            if (self.collection != null && self.collection !== 'undefined') {
                self.clients.couchbaseCollection = self.clients.couchbaseBucket.scope(self.scope).collection(self.collection);
            }
        } else {
            self.clients.couchbaseCollection = self.clients.couchbaseBucket.defaultCollection();
        }
        return self.clients.couchbaseCollection;
    }


    getCouchbaseCollection() {
        var self = this;
        return Promise.try(() => {
            if (self.clients && self.clients.couchbaseBucket && self.clients.couchbaseCollection) {
                console.log('using already open conn');
                return self.clients.couchbaseCollection;
            } else {
                console.log('creating a new conn');
                const result =  this.prepCouchbaseConnection();
                return result;
            }
        }).catch((err) => {
            console.log(err);
            throw err;
        });
    }

    set(key, value, options, cb) {
        var couchbaseStore = this;

        if (typeof options === 'function') {
            cb = options;
            options = {
                ttl: 0
            };
        }

        if (options == null || options === 'undefined') {
            options = {};
        }

        if (cb === undefined) {
            return new Promise(function (resolve, reject) {
                couchbaseStore.set(key, value, options, function (err, result) {
                    err ? reject(err) : resolve(result)
                })
            })
        }

        if (!couchbaseStore.isCacheableValue(value)) {
            return cb(new Error(`"${value}" is not a cacheable value`));
        }

        let ttl = (options.ttl || options.ttl === 0) ? options.ttl : couchbaseStore.ttl;
        ttl = ttl * 60;

        couchbaseStore.getCouchbaseCollection()
            .then((collection)  => {
                 const result =  collection.upsert(key, value, { expiry: ttl });
                 return result;
            })
            .catch((err) => {
                console.log(err);
                cb(err);
            })
    }

    get(key, options, cb) {
        var couchbaseStore = this;

        if (typeof options === 'function') {
            cb = options;
            options = {};
        }

        if (cb === undefined) {
            return new Promise(function (resolve, reject) {
                couchbaseStore.get(key, options, function (err, result) {
                    err ? reject(err) : resolve(result)
                })
            })
        }

        couchbaseStore.getCouchbaseCollection()
            .then((collection) => {
                return collection.get(key);
            }).then((value) => cb(null, value.result))
            .catch((err) => {
                cb(err);
            });
    }

    del(key, options, cb) {
        var couchbaseStore = this;

        if (typeof options === 'function') {
            cb = options;
            options = {};
        }

        if (cb === undefined) {
            return new Promise(function (resolve, reject) {
                couchbaseStore.del(key, options, function (err, result) {
                    err ? reject(err) : resolve(result)
                })
            })
        }

        couchbaseStore.getCouchbaseCollection()
            .then((collection) => {
                return collection.remove(key);
            }).then((value) => cb(null, value))
            .catch((err) => {
                cb(err);
            });
    }

    reset(cb) {
        var couchbaseStore = this;

        if (cb === undefined) {
            return new Promise(function (resolve, reject) {
                couchbaseStore.reset(function (err, result) {
                    err ? reject(err) : resolve(result)
                })
            })
        }

        couchbaseStore.getCouchbaseCollection()
            .then(()=> couchbaseStore.cluster.buckets)
            .then((bucketMgr) => bucketMgr.flushBucket(couchbaseStore.bucket))
            .then((value) => cb(null, value))
            .catch((err) => cb(err));
    }

    keys(pattern, cb) {
        var couchbaseStore = this;

        if (cb === undefined) {
            return new Promise(function (resolve, reject) {
                couchbaseStore.keys(key, options, function (err, result) {
                    err ? reject(err) : resolve(result)
                })
            })
        }

        const bucketReplacement = `\`${storeArgs.bucket}\`.\`${storeArgs.scope}\`.\`${storeArgs.collection}\``;

        const query = `
        SELECT RAW meta().id
        FROM ${bucketReplacement} 
        WHERE REGEXP_CONTAINS(META().id, $PATTERN)
        `;

        const optionsQuery = {
            parameters: {
                PATTERN: pattern,
            }
        };

        couchbaseStore.cluster.query(query, optionsQuery)
            .then((results) => {
                return results.rows;
            })
            .then((rows) => cb(null, rows))
            .catch((err) => cb(err));
    }

    ttl(key, cb) {
        var couchbaseStore = this;

        if (cb === undefined) {
            return new Promise(function (resolve, reject) {
                couchbaseStore.ttl(key, function (err, result) {
                    err ? reject(err) : resolve(result)
                })
            })
        }

        couchbaseStore.getCouchbaseCollection()
            .then((collection) => {
                return collection.get(key);
            })
            .then((result) => {
                return (result.expiryTime - Math.floor(Date.now() / 1000));
            })
            .then((exp) => cb(null, exp))
            .catch((err) => cb(err));
    }
}

exports = module.exports = {
    create: (args) => {
        return new CouchbaseStore(args);
    }
};

index_old.js
const couchbase = require('couchbase');

const couchbaseStore = (...args) => {

  let couchbaseCluster = null;
  let couchbaseBucket = null;
  let couchbaseCollection = null;
  let defaultTtl = 0;

  /* configuration example.. see couchbase docs for more
   * {
   *    connectionString:'',
   *    connectionOptions: {
   *      username:'',
   *      password:''
   *    },
   *    bucket: '',
   *    scope: '',
   *    collection: ''
   * } 
   */

  couchbaseCluster = couchbase.connect(args.connectionString, args.connectionOptions).resolve();
  couchbaseBucket = couchbaseCluster.bucket(args.bucket);

  //defining scope and collection, otherwise falling back on default one.
  if (args.scope != null && args.scope === 'undefined') {
    couchbaseScope = args.scope;

    if (args.collection != null && args.collection === 'undefined') {
      couchbaseCollectionName = args.collection;
      couchbaseCollection = couchbaseBucket.scope(args.scope).collection(args.collection);
    }

  } else {
    couchbaseCollection = bucket.defaultCollection();
    args.scope = '_default';
    args.collection = '_default';
  }

  if (args.ttl != null && args.ttl !== 'undefined') {
    defaultTtl = args.ttl;
  }

  const storeArgs = args;

  let self = {
    name: 'couchbase',
    isCacheableValue: storeArgs.isCacheableValue || (value => value !== undefined && value !== null),
  };

  self.getClient = () => couchbaseBucket;

  self.set = (key, value, options, cb) => {
    if (typeof options === 'function') {
      cb = options;
      options = {};
    }
    options = options || {};

    if (!cb) {
      cb = (err, result) => (err ? reject(err) : resolve(result));
    }

    if (!self.isCacheableValue(value)) {
      return cb(new Error(`"${value}" is not a cacheable value`));
    }

    let ttl = (options.ttl || options.ttl === 0) ? options.ttl : defaultTtl.ttl;
    ttl = ttl * 60;

    return couchbaseCollection.upsert(key, value, { expiry: ttl })
      .then((result) => cb(null, result));
  }

  self.get = (key, options, cb) => {

    if (typeof options === 'function') {
      cb = options;
    }

    if (!cb) {
      cb = (err, result) => (err ? reject(err) : resolve(result));
    }
    return couchbaseCollection.get(key)
      .then((value) => cb(null, value.result));
  };

  self.del = (key, options, cb) => {
    if (typeof options === 'function') {
      cb = options;
    }

    return couchbaseCollection.remove(key)
      .then(() => cb);
  };

  self.reset = (cb) => {
    return couchbaseCluster.buckets()
      .then( (bucketMgr) => bucketMgr.flushBucket(storeArgs.bucket))
      .then(() => cb);
  };

  //to implement with query pagination. needs the following index: CREATE INDEX `all_keys` ON `bucket`((meta().`id`))
  self.keys = (pattern, cb) => {
    if (typeof pattern === 'function') {
      cb = pattern;
      pattern = '%';
    } else {
      pattern = pattern.replaceAll('*', '%');
    }

    const bucketReplacement = `\`${storeArgs.bucket}\`.\`${storeArgs.scope}\`.\`${storeArgs.collection}\``;

    const query = `
      SELECT RAW meta().id
      FROM ${bucketReplacement} 
      WHERE REGEXP_CONTAINS(META().id, $PATTERN)
    `;

    const optionsQuery = {
      parameters: {
        PATTERN: pattern,
        LIMIT: 1,
        OFFSET: 2
      }
    }

    couchbaseCluster.query(query, optionsQuery)
      .then((results) => {
        return results.rows;
      })
      .then((rows) => cb(null, rows));
  };

  //check if it is seconds or minutes
  self.ttl = (key, cb) => couchbaseCollection.get(key)
  .then((result) => {
    return (result.expiryTime - Math.floor(Date.now() / 1000));
  }).then( (secondsToExpiration) => cb(null, secondsToExpiration));

  return self;
};

const methods = {
  create: (...args) => couchbaseStore(...args),
};

module.exports = methods;
